### PR TITLE
Problem: no integration test for transaction index

### LINKED ISSUE #246: Problem: no integration test for transaction index
There are some complex logics around transaction index, it's good to check them in the integration test:
- Try different status of tx, and check how they are handled by json rpc API
  - failed, check error message, check it's not included in getBlock API.
  - reverted, check it is included in getBlock API, check getBlockAndTxIndex API works ok, check the tx index in the logs are correct.

### PR DESCRIPTION
**Pull request by damoncro**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Close #246 

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: integration_tests/test_basic.py [hunk 1]
```diff
@@ -217,3 +218,70 @@ def test_statesync(cronos):
...
+            {
+                "to": ADDRS["community"],
+                "value": 10000,
+                "gasPrice": w3.eth.gas_price,
+                "gas": 1,
+            },
+            KEYS["validator"],
+        )["transactionHash"]
+    except ValueError as error:
+        assert "out of gas" in str(error)
```
Comment (yihuang) [inline integration_tests/test_basic.py:248]: https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions
there's `pytest.raises` to expect error.

Comment (damoncro) [reply integration_tests/test_basic.py:248]: But I need to assert the string.

Comment (damoncro) [reply integration_tests/test_basic.py:248]: Oh raises also ok to assert inside?

Comment (damoncro) [reply integration_tests/test_basic.py:248]: I'll check

File: integration_tests/utils.py [hunk 2]
```diff
@@ -441,3 +441,83 @@ def call_contact(self, w3):
         assert "world" == self.contract.functions.greet().call()
 
         return send_store_tx
+
+
+class TestRevert:
```
Comment (yihuang) [inline integration_tests/utils.py:446]: It seems many codes can be reused with `class Greeter`.

Comment (damoncro) [reply integration_tests/utils.py:446]: Yes. This can be improved...

File: integration_tests/test_basic.py (hunk 1 above)
Comment (yihuang) [reply integration_tests/test_basic.py:248]: yes, `with pytest.raises(...) as exc_info` and assert the `exc_info`.

**Comment by yihuang** [comment]
- try to include multiple txs in same block, and check the indexes.
- include a cosmos level failure case, and check it doesn't appear in `get_block` response.
- check the tx_index and log_index in tx receipt are correct.
- check `get_transaction_by_block` works.

File: integration_tests/utils.py [hunk 3]
```diff
@@ -109,6 +111,12 @@ def wait_for_ipc(path, timeout=40.0):
             )
 
 
+def wait_for_current_block_confirmation(w3):
```
Comment (yihuang) [inline integration_tests/utils.py:114]: This method seems like same as `wait_for_new_blocks(1)`.

**Review by yihuang — APPROVED** [review]

**Comment by damoncro** [comment]
Sorry, the PR is still under polished. Please hold it first.

File: integration_tests/utils.py [hunk 4]
```diff
@@ -383,61 +388,143 @@ def __init__(self, private_key=KEYS["validator"], chain_id=777):
...
             )
 
-            signed_txn = w3.eth.account.sign_transaction(
-                transaction, private_key=self.private_key
-            )
-
-            # print("Deploying contract...")
-            # Send this signed transaction
-            self._contract_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
+            tx_hash, tx_receipt = self.sign(transaction).send().wait()
```
Comment (yihuang) [inline integration_tests/utils.py:404]: I think we can use `send_transaction(tx)` here?

File: integration_tests/utils.py (hunk 3 above)
Comment (damoncro) [reply integration_tests/utils.py:114]: It is different, sometimes the block number did not increase when the first time to call ` w3.eth.get_block_number()`. The purpose here is not simply wait for one block.

File: integration_tests/utils.py (hunk 4 above)
Comment (damoncro) [reply integration_tests/utils.py]: let me replace with send_transaction.

File: integration_tests/utils.py (hunk 3 above)
Comment (yihuang) [reply integration_tests/utils.py:114]: I don't see the difference, isn't the code is just waiting for a new block mined?

Comment (damoncro) [reply integration_tests/utils.py]: Sometimes, I found deploy contracts thread and call contracts thread will make all transactions included in one block. With this, it make sure they will split it into two or three blocks (1 empty).

But now, I can not find it is necessary anymore, I already deleted the `wait_for_current_block_confirmation`.

File: integration_tests/test_basic.py [hunk 5]
```diff
@@ -217,3 +223,176 @@ def test_statesync(cronos):
...
+
+
+def assert_receipt_transaction_and_block(w3, futures):
+    receipts = []
+    for future in concurrent.futures.as_completed(futures):
+        # name = future_to_contract[future]
+        try:
+            data = future.result()
+        except Exception as exc:
+            print("%s" % (exc))
```
Comment (yihuang) [inline integration_tests/test_basic.py:363]: why ignore the exception here?

File: integration_tests/test_basic.py [hunk 6]
```diff
@@ -217,3 +223,176 @@ def test_statesync(cronos):
...
+    for receipt in receipts:
+        # check in the same block
+        assert receipt["blockNumber"] == block_number
+        # check transactionIndex
+        transaction_index = receipt["transactionIndex"]
+        assert transaction_index in tx_indexes
+        tx_indexes.remove(transaction_index)
+
+    block = w3.eth.get_block(block_number)
+    print(block)
```
Comment (yihuang) [inline integration_tests/test_basic.py:380]: remove debug prints?

**Review by yihuang — APPROVED** [review]
LGTM, just minor suggestions:
- exception handling
- cleanup debug prints

**Pull request merged**